### PR TITLE
[MPS] Move max_pool2d to mps dispatch key

### PR DESCRIPTION
**Pull request by qqaatw**
Related issue: #77394

This PR also modifies some assertions in the codegen, an explanatory comment for it has been added.

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2138,8 +2138,8 @@
   input, weight, bias: linear_backward(input, grad, weight, grad_input_mask)
 
 #mps
-- name: _mps_max_pool2d(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> Tensor
-  self: mps_max_pool2d_backward(grad, self, kernel_size, stride, padding, dilation, ceil_mode)
+- name: max_pool2d(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> Tensor
+  self: max_pool2d_backward(grad, self, kernel_size, stride, padding, dilation, ceil_mode)
```
Comment (malfet) [inline tools/autograd/derivatives.yaml:2142]: This entry should already be declared in the derivaties.yaml , shouldn't' it?

**Review by malfet — COMMENTED** [review]
Looks fine, but please fix build and test first

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (qqaatw) [reply tools/autograd/derivatives.yaml:2174]: @malfet Hello, could you please give some suggestions regarding the CI failures?

**Review by albanD — APPROVED** [review]
Perfect! Thanks for the update to the codegen!

**Comment by kulinseth** [comment]
@qqaatw , I am seeing error in torchbench:

```
python run.py maml_omniglot -d mps -m eager -t train
```
```
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: derivative for aten::max_pool2d_backward is not implemented
```

**Comment by qqaatw** [comment]
@kulinseth I will take a look at it soon.

**Comment by kulinseth** [comment]
> @kulinseth I will take a look at it soon.

@qqaatw , even with the revert, its the same issue. I think its something to do with higher order gradients not working with mps specific function registered.

**Comment by qqaatw** [comment]
@kulinseth Yeah indeed. I remember that the higher order gradient of `linear` on mps has this issue as well.

**Pull request closed without merging**